contracts/debt/IncurDebtV1.sol

```diff
@@ -0,0 +1,453 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity ^0.8.10;
+
+import "../interfaces/IOHM.sol";
+import "../interfaces/IsOHM.sol";
+import "../interfaces/IgOHM.sol";
+import "../interfaces/IERC20.sol";
+import "../interfaces/IStaking.sol";
+import "./interfaces/ITreasury.sol";
+import "../libraries/SafeERC20.sol";
+import "../types/OlympusAccessControlledV2.sol";
+
+error IncurDebtV1_NotBorrower(address _borrower);
+error IncurDebtV1_InvaildNumber(uint256 _amount);
+error IncurDebtV1_WrongTokenAddress(address _token);
+error IncurDebtV1_AlreadyBorrower(address _borrower);
+error IncurDebtV1_AboveGlobalDebtLimit(uint256 _limit);
+error IncurDebtV1_AboveBorrowersDebtLimit(uint256 _limit);
+error IncurDebtV1_LimitBelowOutstandingDebt(uint256 _limit);
+error IncurDebtV1_AmountAboveBorrowerBalance(uint256 _amount);
+error IncurDebtV1_AmountMoreThanBorrowersLimit(uint256 _borrower);
+error IncurDebtV1_OHMAmountMoreThanAvailableLoan(uint256 _amount);
+error IncurDebtV1_BorrowerHasNoOutstandingDebt(address _borrower);
+error IncurDebtV1_BorrowerStillHasOutstandingDebt(address _borrower);
+
+contract IncurDebtV1 is OlympusAccessControlledV2 {
+    using SafeERC20 for IERC20;
+
+    event GlobalLimit(uint256 indexed _limit);
+    event BorrowerAllowed(address indexed _borrower);
+    event BorrowerRevoked(address indexed _borrower);
+    event BorrowerDebtLimitSet(address indexed _borrower, uint256 indexed _limit);
+    event CollateralInSOHMIncreased(address indexed _borrower, uint256 indexed _currentCollateralInSOHM);
+    event BorrowerDeposit(
+        address indexed _borrower,
+        uint256 indexed _amountToDeposit,
+        address indexed _tokenToDepositFundsWith
+    );
+    event Borrowed(
+        address indexed _borrower,
+        uint256 _amountToBorrow,
+        uint256 indexed _borrowersDebt,
+        uint256 indexed _totalOutstandingGlobalDebt
+    );
+    event DebtPaidWithOHM(
+        address indexed _borrower,
+        uint256 _paidDebt,
+        uint256 indexed _currentDebt,
+        uint256 indexed _totalOutstandingGlobalDebt
+    );
+    event Withdrawal(
+        address indexed _borrower,
+        uint256 _amountToWithdraw,
+        address _tokenToWithdrawFundsWith,
+        address indexed _receiver,
+        uint256 indexed _collateralInSOHMLeft
+    );
+    event DebtPaidWithCollateral(
+        address indexed _borrower,
+        uint256 indexed _paidDebt,
+        uint256 _currentCollateralInSOHM,
+        uint256 indexed _currentDebt,
+        uint256 _totalOutstandingGlobalDebt
+    );
+    event DebtPaidWithCollateralAndBurnTheRest(
```

Comment:
Do we really need this level of granularity? Shouldn't DebtPaidWithCollateral() be enough for analytic reasons.

Reply:
I think to be on the safe side its best we have all information well detailed and pinpointed where values are emitted from.

This `DebtPaidWithCollateralAndBurnTheRest` is different from this `DebtPaidWithCollateral` in the sense that the latter only pays your debt will the other pays debts and the burnt the rest, will be important to know the difference when listening for them.